R/get_mol_kegg.R

```diff
@@ -0,0 +1,113 @@
+utils::globalVariables(".data")
+#' Download compound .mol files from KEGG
+#'
+#' Downloads mol files corresponding to individual compounds or compounds in a
+#' pathway from KEGG.
+#'
+#' @param compound_ids character vector of KEGG compound IDs---5 digits
+#'   prepended with a "C".
+#' @param pathway_ids character vector of KEGG pathway IDs---5 digits prepended
+#'   with "map".
+#' @param dir path to a folder to save mol files in.
+#'
+#' @returns a tibble with the columns `compound_ids`, `pathway_ids` (if used),
+#'   and `mol_paths` (paths to downloaded .mol files)
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#' get_mol_kegg(compound_ids = c("C16181", "C06074"), dir = tempdir())
+#' get_mol_kegg(pathway_ids = "map00253", dir = tempdir())
+#' }
+get_mol_kegg <- function(compound_ids, pathway_ids, dir){
+  
+  #TODO: implement redownload=FALSE arg
+  
+  if(missing(dir)) stop("`dir` is required")
+  if ((missing(compound_ids) & missing(pathway_ids)) |
+      !missing(compound_ids) & !missing(pathway_ids)) {
+    stop("One of `compound_id` or `pathway_id` are required")
+  }
+  
+  #if compounds are provided
+  if (!missing(compound_ids)) {
+    if (!all(stringr::str_detect(compound_ids, "^[C][:digit:]{5}$"))) {
+      stop("Some compound_ids are not in the correct KEGG format")
+    }
+    fs::dir_create(dir)
+    out_tbl <-
+      tibble::tibble(compound_id = compound_ids) %>% 
+      dplyr::mutate(mol_path = fs::path(dir, .data$compound_id, ext = "mol"))
+  }
+  # if pathways are provided
+  if (!missing(pathway_ids)) {
+    if (!all(stringr::str_detect(pathway_ids, "^[m][a][p][:digit:]{5}$"))) {
+      stop("Some pathway_ids are not in the correct KEGG format")
+    }
+    fs::dir_create(dir, pathway_ids)
+    compound_ids_list <- lapply(pathway_ids, keggGetCompounds)
+    names(compound_ids_list) <- pathway_ids
+    out_tbl <- 
+      tibble::enframe(compound_ids_list, name = "pathway_id", value = "compound_id") %>% 
+      tidyr::unnest(tidyselect::everything()) %>% 
+      dplyr::mutate(mol_path = fs::path(dir, .data$pathway_id, .data$compound_id, ext = "mol"))
+    
+  }
+  
+  # Download mols
+  .get_mol_kegg <- function(compound_id) {
```

Comment:
@Aariq could you explain to me what the function name with the period before it is doing?

Reply:
The function `.get_mol_kegg()` gets a .mol file for a single `compound_id`.  The enclosing `get_mol_kegg()` function vectorizes this (line 73).  Can't remember where, but I've seen this pattern before in R packages that are like:

``` r
foo <- function(vec) {
  .foo <- function(x){
    ...
  }
  lapply(vec, .foo)
}
```

But `.get_mol_kegg` could be called anything.